core/store/src/trie/raw_node.rs

```diff
@@ -0,0 +1,163 @@
+        &mut self.0[usize::from(index)]
+    }
+}
+
+impl<T: BorshSerialize> BorshSerialize for Children<T> {
+    fn serialize<W: std::io::Write>(&self, wr: &mut W) -> std::io::Result<()> {
+        let mut bitmap: u16 = 0;
+        let mut pos: u16 = 1;
+        for child in self.0.iter() {
+            if child.is_some() {
+                bitmap |= pos
+            }
+            pos <<= 1;
+        }
+        bitmap.serialize(wr)?;
+        self.0.iter().flat_map(Option::as_ref).map(|child| child.serialize(wr)).collect()
+    }
+}
+
+impl<T: BorshDeserialize> BorshDeserialize for Children<T> {
+    fn deserialize_reader<R: std::io::Read>(rd: &mut R) -> std::io::Result<Self> {
+        let mut bitmap = u16::deserialize_reader(rd)?;
+        let mut children = Self::default();
+        while bitmap != 0 {
+            let idx = bitmap.trailing_zeros() as u8;
+            bitmap &= bitmap - 1;
+            children[idx] = Some(T::deserialize_reader(rd)?);
+        }
+        Ok(children)
+    }
+}
+
+mod children {
+    struct Debug<'a, T>((u8, &'a T));
+
+    impl<T: std::fmt::Debug> std::fmt::Debug for Debug<'_, T> {
+        fn fmt(&self, fmtr: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(fmtr, "{}: {:?}", self.0 .0, self.0 .1)
+        }
+    }
+
+    impl<T: std::fmt::Debug> std::fmt::Debug for super::Children<T> {
+        fn fmt(&self, fmtr: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            fmtr.debug_list().entries(self.iter().map(Debug)).finish()
+        }
+    }
+}
+
+#[test]
+fn test_encode_decode() {
+    fn test(node: RawTrieNode, encoded: &[u8]) {
+        let node = RawTrieNodeWithSize { node, memory_usage: 42 };
+        let mut buf = node.try_to_vec().unwrap();
+        assert_eq!(encoded, buf.as_slice());
+        assert_eq!(node, RawTrieNodeWithSize::try_from_slice(&buf).unwrap());
+
+        // Test that adding garbage at the end fails decoding.
+        buf.push(b'!');
+        let got = RawTrieNodeWithSize::try_from_slice(&buf);
+        assert!(got.is_err(), "got: {got:?}");
+    }
+
+    let value = ValueRef { length: 3, hash: CryptoHash::hash_bytes(&[123, 245, 255]) };
+    let node = RawTrieNode::Leaf(vec![1, 2, 3], value.clone());
+    let encoded = [
+        0, 3, 0, 0, 0, 1, 2, 3, 3, 0, 0, 0, 194, 40, 8, 24, 64, 219, 69, 132, 86, 52, 110, 175, 57,
+        198, 165, 200, 83, 237, 211, 11, 194, 83, 251, 33, 145, 138, 234, 226, 7, 242, 186, 73, 42,
+        0, 0, 0, 0, 0, 0, 0,
+    ];
+    test(node, &encoded);
+
+    let mut children = Children::default();
+    children[3] = Some(super::Trie::EMPTY_ROOT);
+    let node = RawTrieNode::BranchNoValue(children);
+    let encoded = [
+        1, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 42, 0, 0, 0, 0, 0, 0, 0,
+    ];
+    test(node, &encoded);
+
+    let mut children = Children::default();
+    children[3] = Some(super::Trie::EMPTY_ROOT);
```

Comment:
No issue with your code, but this test seemed a bit lacking - I think we should at least test for (1) multiple children, (2) first children set, (3) last children set

Reply:
Done.